engines/sci/engine/kvideo.cpp

```diff
@@ -170,37 +201,89 @@ reg_t kShowMovie(EngineState *s, int argc, reg_t *argv) {
 		// The first parameter is a subop. Some of the subops set the accumulator.
 		// The interpreter implements subops 0-6. KQ6 only calls 0, 1, 2, 3, 6.
-		// Subop 0 plays the AVI; it is the only one that needs to be implemented.
 
 		switch (argv[0].toUint16()) {
 		case 0: {
-			Common::String filename = s->_segMan->getString(argv[1]);
+			_movieConf = new MovieConfig();
+			_movieConf->filename = s->_segMan->getString(argv[1]);
+			retval = NULL_REG;
+			break;
+		}
+		case 1:
+			if (_movieConf) {
+				int16 *dest[4] = { &_movieConf->x, &_movieConf->y, &_movieConf->w, &_movieConf->h };
+				for (int i = 1; i < argc; ++i)
+					*dest[i - 1] = argv[i].toSint16();
+			}
+			retval = NULL_REG;
+			break;
+		case 2:
+			if (_movieConf) {
+				switch (argc) {
+				case 5:
+					_movieConf->playTo = argv[4].toUint16();
+					// fallthrough
+				case 4:
+					_movieConf->playFrom = argv[3].toUint16();
+					// fallthrough
+				case 3:
+					_movieConf->pauseToken = argv[2];
+					// fallthrough
+				case 2:
+					_movieConf->mode = argv[1].toUint16();
+					break;
+				default:
+					break;
+				}
+			}
+
+			// For KQ6, this changes the vertical 200/440 upscaling to 200/400, since this is the expected behavior. Also,
+			// the calculation of the scaled x/y coordinates works slightly differently compared to the normal gfx rendering.
+			g_sci->_gfxScreen->gfxDriver()->setFlags(GfxDriver::kMovieMode);
+
 			videoDecoder.reset(new Video::AVIDecoder());
-			if (!videoDecoder->loadFile(filename.c_str())) {
-				warning("Failed to open movie file %s", filename.c_str());
+			if (!videoDecoder->loadFile(_movieConf->filename.c_str())) {
+				warning("Failed to open movie file %s", _movieConf->filename.c_str());
 				videoDecoder.reset();
 			}
-			retval = TRUE_REG;
+
+			syncLastFrame = false;
+			retval = NULL_REG;
+
 			break;
-		}
 		default:
+			// This will trigger on case 6 which is the close/deinit function. We don't need that.
 			debug(kDebugLevelVideo, "Unhandled kShowMovie subop %d", argv[0].toUint16());
 		}
 	}
 
 	if (videoDecoder) {
-		bool is8bit = videoDecoder->getPixelFormat().bytesPerPixel == 1;
+		if (videoDecoder->getPixelFormat().bytesPerPixel > 1)
+			syncLastFrame = false;
 
-		playVideo(*videoDecoder);
+		playVideo(*videoDecoder, _movieConf);
 
 		// HACK: Switch back to 8bpp if we played a true color video.
 		// We also won't be copying the screen to the SCI screen...
 		if (switchedGraphicsMode)
 			g_sci->_gfxScreen->gfxDriver()->initScreen();
-		else if (is8bit) {
+		else if (syncLastFrame) {
 			g_sci->_gfxScreen->kernelSyncWithFramebuffer();
 			g_sci->_gfxPalette16->kernelSyncScreenPalette();
 		}
+
+		g_sci->_gfxScreen->gfxDriver()->clearFlags(GfxDriver::kMovieMode);
+
+		if (_movieConf) {
+			// Unfreeze the engine. KQ6 Windows does not run the video modally, since it happens via OS functions.
+			// So the engine is put into a waiting state until signalled to continue. The orignal interpreter does
```

Comment:
"original"

Reply:
We have someone who's actually reading the comments! :) Also, since we're using American English, it should be "signaled"  instead of "signalled" (yes, I know, I'm nitpicking).

https://sapling.ai/usage/signaled-vs-signalled

```suggestion
			// So the engine is put into a waiting state until signaled to continue. The original interpreter does
```

Reply:
@blugr I was hoping you'd rip into me for this, and rightly so!